lua-api-crates/filesystem/src/lib.rs

```diff
@@ -29,6 +33,63 @@ async fn read_dir<'lua>(_: &'lua Lua, path: String) -> mlua::Result<Vec<String>>
     Ok(entries)
 }
 
+// similar (but not equal) to the shell command basename
+async fn basename<'lua>(_: &'lua Lua, path: String) -> mlua::Result<String> {
+    // to check if the path actually exists, we can use:
+    /* let path = smol::fs::canonicalize(path)
+    .await
+    .map_err(mlua::Error::external)?; */
+    let path = Path::new(&path);
+    if let Some(basename) = path.file_name() {
+        if let Some(utf8) = basename.to_str() {
+            Ok(utf8.to_string())
+        } else {
+            return Err(mlua::Error::external(anyhow!(
+                "path entry {} is not representable as utf8",
+                path.display()
+            )));
+        }
+    } else {
+        // file_name returns None if the path name ends in ..
+        Ok("..".to_string())
+    }
+}
+
+// return the path without its final component if there is one
+// similar to the shell command dirname
+async fn dirname<'lua>(_: &'lua Lua, path: String) -> mlua::Result<String> {
+    let path = Path::new(&path);
+    if let Some(parent_path) = path.parent() {
+        if let Some(utf8) = parent_path.to_str() {
+            Ok(utf8.to_string())
+        } else {
+            return Err(mlua::Error::external(anyhow!(
+                "path entry {} is not representable as utf8",
+                path.display()
+            )));
+        }
+    } else {
+        // parent returns None if the path terminates in a root or prefix
+        Ok("".to_string())
```

Comment:
I'm not sure about this one.
`dirname /` produces `/` in the shell.

Perhaps we should return the input path here?

I think the empty string is not a good return value here because it can be interpreted differently between unix and windows in different contexts.

Reply:
I did consider returning `/`, but I am not sure if that would make sense in windows? (Basically I just know unix systems in general.)

Reply:
Or should we return whatever corresponds to root in windows? I can probably make that work too.

Reply:
What I suspect is that we only hit this case when the input is a root, so we can simply return the input string in this case?

Reply:
I have changed it so it returns the input for now.